### PR TITLE
New goto-analyzer infrastructure : take 2

### PR DESCRIPTION
**Pull request by martin-cs**
This is a refactoring of Lucas' aisimplifier branch along with a number of other features. It changes the way goto-analyze works so that you can independently pick the task (show domain, verify assertions, simplify), the type of analysis (flow-sensitive at the moment), the domain (constants and intervals) and the output (text, JSON, XML).

### CODE REVIEW

**Comment by tautschnig** [comment]
Is there any urgency for getting this one merged? Otherwise it would be a really useful pull request to study in detail on 28th November.

**Comment by kroening** [comment]
This is too invasive for the release today; in particular, various industrial users would have to re-run their tests.

**Comment by martin-cs** [comment]
On Fri, 2016-11-18 at 01:37 -0800, Daniel Kroening wrote:
> This is too invasive for the release today; in particular, various industrial users would have to re-run their tests.

Do you want any specific changes for it to be mergeable in the future?

Cheers,
 - Martin

**Comment by martin-cs** [comment]
On Thu, 2016-11-17 at 14:43 -0800, Michael Tautschnig wrote:
> Is there any urgency for getting this one merged?

It is the basis for several of the patch sets we are working on for
toyota, so it would save quite a bit of hassle and allow us faster
release cycles.  I doubt it will be merged in the next 10 days though.

>  Otherwise it would be a really useful pull request to study in detail on 28th November.

:-\  As in "here is an example of what not to do"...

Cheers,
 - Martin

**Comment by tautschnig** [comment]
More like: "here are many examples of how we could make reviewing of pull requests much easier." Sorry if I'm breathing too much Amazonian air, which is all about learning, yet never about blaming.

File: src/goto-analyzer/static_show_domain.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+\*******************************************************************/
+
+bool static_show_domain(
+  const goto_modelt &goto_model,
+  const optionst &options,
+  message_handlert &message_handler,
+  std::ostream &out)
+{
+  ai_baset *domain = NULL;
```
Comment (danpoe) [inline src/goto-analyzer/static_show_domain.cpp]: The variable does not represent a domain. How about naming it ``ai``?

**Comment by danpoe** [comment]
There's a few aspects I could see in which the code does not follow the coding guidelines yet: whitespace around operators, maximum line length, contains tabs.

**Comment by danpoe** [comment]
The constant propagator included here has a few issues that have been fixed in a different branch over some time. @martin-cs, @tautschnig, @kroening  Should I create a new branch for those fixes and then create a second pull request on top of this one, or should we cherry-pick these commits and add them to this pull request?

File: regression/goto-analyzer/Makefile [hunk 2]
```diff
@@ -12,3 +12,9 @@ show:
 			vim -o "$$dir/*.java" "$$dir/*.out"; \
 		fi; \
 	done;
+
+clean:
+	find . -name *.*~ | xargs rm -f
+	find . -name *.out | xargs rm -f
+	find . -name *.goto | xargs rm -f
+	rm -f tests.log
```
Comment (tautschnig) [inline regression/goto-analyzer/Makefile]: While at it: one could use `find ... -delete` instead of `find ... | xargs rm -f`

**Comment by tautschnig** [comment]
@danpoe I might just be missing a bit of information here: how do the branches relate that you are referring to? Have you worked off of some common ancestor? Maybe you can just point to the two branch heads that you are referring to.

As we haven't found enough time to actually apply the "theory" to this pull request: Overall, I believe this pull request should be split into several, smaller pull requests as there seems to be a lot of preparation in place for the "hightlights" of this pull request. Those smaller ones should then get reviewed as quickly as possible so as to enable the final, intrusive pull request.

**Comment by danpoe** [comment]
@tautschnig I have a branch danpoe:unwind-bounds which is a branch of martin-cs:goto-analyzer-2 (this pull request). The unwind-bounds branch contains some fixes to the constant propagator included here (but unfortunately not nicely separated into commits that only affect the constant propagator).

**Comment by tautschnig** [comment]
@danpoe: Here's what I would do (yes, this usually takes quite a bit of time):
1. Just working inside your branch, squash commits so that your branch (on top of martin-cs:goto-analyzer-2) is a clean series of commits that in itself would make a good pull request.
2. Re-order the commits so that those affecting the constant propagator are at the top (via `git rebase -i`).
3. Create a new branch that just has those commits (either via cherry-pick or just creating a branch at the last commit id).
4. Create a pull request in martin-cs' fork.
5. Have @martin-cs merge your pull request into the branch, at which point those commits will immediately show up in here.
(@peterschrammel and I have done this dance more than once already, it works quite well I believe.)

**Comment by martin-cs** [comment]
On Mon, 2016-12-05 at 06:10 -0800, Michael Tautschnig wrote:
> @danpoe: Here's what I would do (yes, this usually takes quite a bit of time):
> 1. Just working inside your branch, squash commits so that your branch (on top of martin-cs:goto-analyzer-2) is a clean series of commits that in itself would make a good pull request.
> 2. Re-order the commits so that those affecting the constant propagator are at the top (via `git rebase -i`).
> 3. Create a new branch that just has those commits (either via cherry-pick or just creating a branch at the last commit id).
> 4. Create a pull request in martin-cs' fork.
> 5. Have @martin-cs merge your pull request into the branch, at which point those commits will immediately show up in here.
> (@peterschrammel and I have done this dance more than once already, it works quite well I believe.)

I need to refactor martin-cs:goto-analyzer-2 as it no longer easily
rebases onto master.  I simply have not had the time.

I would be OK with @danpoe pulling out all of the fixes to the constant
domain and getting these merged to trunk as an independent set of
patches.  This is, ofcourse, if we can get it merged quickly as doing
this will hold up development.

Cheers,
 - Martin

**Comment by tautschnig** [comment]
@danpoe @martin-cs Taking this into account, my past approaches to such situations were: before rebasing onto current master, build up one single, long patch series. That would be the case of @danpoe doing all of the steps in my previous post as @martin-cs would then have the full patch series at hand. The next step is to squash that patch series so that a smaller number of individually meaningful commits remain. Usually that's a matter of applying `git rebase -i` until happy. Finally either do the rebase onto current master in that branch, or else create new branches that are made from cherry-picks of that current branch. (In case it helps: I've presently got a 170+ and a 50+ patch series that I'm maintaining in this way.)

**Comment by danpoe** [comment]
@tautschnig Thanks for the detailed explanations! @martin-cs and I will talk in person next week about how we'll address this.

**Comment by martin-cs** [comment]
Reworked and rebased, see:
https://github.com/diffblue/cbmc/pull/424

**Pull request closed without merging**